client/src/main/java/org/eclipse/hono/client/Command.java

```diff
@@ -327,4 +330,11 @@ public String toString() {
             return "Invalid Command";
         }
     }
+
+    private static String processForDeviceId(final String replyToId, final String deviceId) {
+        if (replyToId.startsWith(deviceId + "/")) {
+            return replyToId.replaceFirst(deviceId + "/", deviceId + "/0");
```

Comment:
it feels strange to me that we mark up the *next* segment in order to indicate whether the deviceId needs to be removed or not. Can we instead prepend the deviceId with a `o` or `1`? That would also be more in line with the way we do it for the requestId ...

Reply:
> it feels strange to me that we mark up the _next_ segment in order to indicate whether the deviceId needs to be removed or not. Can we instead prepend the deviceId with a `o` or `1`? That would also be more in line with the way we do it for the requestId ...

@sophokles73 It also uses similar approach as of the _requestId_ for the _deviceId_ handling. In case of _requestId_ no _deviceId_ is added the _replyTo_ address. The _requestId_ is prepended with `1` or `0` to indicate if the _deviceId_ is to be later in the `CommandResponse` class added or not. But for the current problem, the _deviceId_ should be sent with the _replyTo_ address.

In this PR, the _replyTo_ address is checked for the _deviceId_ . If it is present, then the _replyId_ is prepended with `0` to indicate that later int the `CommandResponse` class the _deviceId_ should not be removed. If not present, then the _device-id_ is included in the _replyTo_ address and the _replyId_ is prepended with `1` to indicate the _deviceId_ to be removed later in the below example.
```
control/Tenant/Device/xyz    ->    control/Tenant/Device/0xyz 
control/Tenant/xyz   ->    control/Tenant/Device/1xyz
```

Prepending to the _deviceId_ instead to the _replyId_ would appear as below.
```
control/Tenant/Device/xyz    ->    control/Tenant/0Device/xyz 
control/Tenant/xyz     ->    control/Tenant/1Device/xyz
```
Also an advantage of prepending to the _replyId_ instead to the _deviceId_ is that the `ResourceIdentifier.getResourceId()` returns a valid _deviceId_ instead starting with `1` or `0`. For example `AMQPContext`  uses `getResourceId() `  to set the _deviceId_ in case of unauthenicated devices. IMHO its a matter of  IMHO its a matter of perspective, if the _deviceId_ being added to the _replyId_ or to the _replyTo_ address.